File: utils/CanonicalSystem/canClock.m
classdef canClock 
    methods (Access = public)
        % Canonical Clock constructor.
        % @param[in] scale : Parameter to define the scaling of the canonical system.
        % @param[in] can_clock_index : Flag to define whether the exponential canonical clock(0) should be used or the linear (~0)
        function this = canClock(scale,can_clock_index)
            if (nargin < 2)
                this.index = 1;
            else
                this.index = can_clock_index;
            end
            this.tau = scale;
            this.ac = 1;
        end
        
        %  @param[in] t : Moment on which the desired phase value should be calculated.
        function x = getPhase(this,t)
            if this.index
                x = (1/this.tau)*t;
            else
                x = exp(-(t*this.ac)/this.tau);
            end
        end

        %  @param[in] t : Moment on which the desired phase_dot value should be calculated.
        function x_dot = getPhaseDot(this,t)
            timesteps = length(t);
            if this.index
                x_dot = (1/this.tau)*ones(length(timesteps),1);
            else
                x_dot = -(this.ac/this.tau)*exp(-(t*this.ac)/(this.tau));
            end
        end

        %  @param[in] t : Moment on which the desired phase_ddot value should be calculated.
        function x_ddot = getPhaseDDot(this,t)
            timesteps = length(t);
            if this.index
                x_ddot = zeros(length(timesteps),1);
            else
                x_ddot = (this.ac^2/(this.tau^2))*exp(-(t*this.ac)/(this.tau));
            end
        end

        % Getter.
        function tau = get_tau(this)
            tau = this.tau;
        end

    end

    properties (Access = private)
        index % Flag to choose between desired canonical clock
        tau % Scaling parameter
        ac % Parameter for exponential canonical clock
    end
end

File: experiment/canclock.cpp
#include "canclock.h"

#include <cmath>
#include <armadillo>

/******** public ********/
canclock::canclock(double scale, bool can_clock_index)
{
    this->index = can_clock_index;
    this->tau = scale;
    this->ac = 1;
}

canclock::canclock()
{
    this->index = 1;
    this->tau = 10;
    this->ac = 1;
}

arma::vec canclock::getPhase(arma::vec t)
{
    arma::vec x = arma::vec(t.n_rows);
    if (this->index){x = (1/this->tau)*t;}
    else {x = exp(-(t*this->ac)/this->tau);}
    return x;
}

arma::vec canclock::getPhaseDot(arma::vec t)
{
    arma::vec dx = arma::vec(t.n_rows);
    if (this->index){dx = (1/this->tau)*arma::ones(t.n_rows);}
    else {dx = -(this->ac/this->tau)*exp(-(t*this->ac)/this->tau);}
    return dx;
}

arma::vec canclock::getPhaseDDot(arma::vec t)
{
    arma::vec ddx = arma::vec(t.n_rows);
    if (this->index){ddx.zeros();}
    else {ddx = +(pow(this->ac,2)/pow(this->tau,2))*exp(-(t*this->ac)/this->tau);}
    return ddx;
}

double canclock::get_tau()
{
    return this->tau;
}

bool canclock::get_index()
{
    return this->index;
}


